File: 0238-Product-of-Array-Except-Self.py
class Solution(object):
    def productExceptSelf(self, nums):
        n = len(nums)
        
        finalProduct = [1] * n
        
        for i in range(n-1):
            finalProduct[i + 1] = finalProduct[i] * nums[i]
            
        rightProduct = 1
        for i in range(n-1, -1 , -1):
            finalProduct[i] *= rightProduct
            rightProduct *= nums[i]
            
        return finalProduct

File: 0238-Product-of-Array-Except-Self.js
var productExceptSelf = function(nums) {
    let n = nums.length;
    let finalProd = new Array(n).fill(1);
    let leftProd = 1;
    let rightProd = 1;
    
    for (let i = 0; i < n; i++) {
        finalProd[i] = finalProd[i] * leftProd;
        leftProd = leftProd * nums[i];
        finalProd[n - 1 - i] = finalProd[n - 1 - i] * rightProd;
        rightProd = rightProd * nums[n - 1 - i]
    }
    
    return finalProd;
};
